addons/controllers/clusterbootstrap_controller.go

```diff
@@ -664,7 +1031,64 @@ func (r *ClusterBootstrapReconciler) watchProvider(providerRef *corev1.TypedLoca
 	)
 }
 
-func (r *ClusterBootstrapReconciler) getCNI(
+// GetDataValueSecretNameFromBootstrapPackage attempts to get the data value secret name associated with a ClusterBootstrapPackage.
+// Users have three ways to provide the data values for a ClusterBootstrapPackage: [Inline, SecretRef, ProviderRef], or
+// leave ClusterBootstrapPackage.ValuesFrom field as nil. If data values are provided by ProviderRef, the corresponding
+// controller needs to generate the secret object.
+//
+// Returns:
+// - string: The secret name which references to the Secret CR on mgmt cluster under a particular cluster namespace.
+// - error: whether there is error when getting the secret name.
+func (r *ClusterBootstrapReconciler) GetDataValueSecretNameFromBootstrapPackage(cbPkg *runtanzuv1alpha3.ClusterBootstrapPackage, clusterNamespace string) (string, error) {
+	// When valuesFrom is nil, we interpret it as no data values are needed for the package installation.
+	if cbPkg.ValuesFrom == nil {
+		r.Log.Info(fmt.Sprintf("no data values are provided to the ClusterBootstrapPackage.ValuesFrom field. ClusterBootstrapPackage.RefName: %s", cbPkg.RefName))
+		return "", nil
+	}
+
+	// TODO: Handle inline valueFrom. https://github.com/vmware-tanzu/tanzu-framework/issues/1694
+
+	if cbPkg.ValuesFrom.SecretRef != "" {
+		return cbPkg.ValuesFrom.SecretRef, nil
+	}
+
+	if cbPkg.ValuesFrom.ProviderRef != nil {
+		gvr, err := r.getGVR(schema.GroupKind{Group: *cbPkg.ValuesFrom.ProviderRef.APIGroup, Kind: cbPkg.ValuesFrom.ProviderRef.Kind})
+		if err != nil {
+			r.Log.Error(err, "unable to get GVR")
+			return "", err
+		}
+		provider, err := r.dynamicClient.Resource(*gvr).Namespace(clusterNamespace).Get(r.context, cbPkg.ValuesFrom.ProviderRef.Name, metav1.GetOptions{}, "status")
+		if err != nil {
+			r.Log.Error(err, "unable to fetch provider", "GVR", gvr)
+			return "", err
+		}
+		secretName, found, err := unstructured.NestedString(provider.UnstructuredContent(), "status", "secretRef")
+		if err != nil {
+			r.Log.Error(err, "unable to fetch secretRef in provider", "GVR", gvr)
+			return "", err
+		}
+		if !found {
+			// In this case, we expect the secretRef to be present under status subresource and its value gets updated by
+			// the corresponding controller. However, the config controller might not create the secret in time.
+			r.Log.Info("provider status does not have secretRef", "GVR", gvr)
+			return "", nil
+		}
+		return secretName, nil
+	}
+
+	// When valuesFrom is not nil, but either valuesFrom.Inline, valuesFrom.SecretRef, or valuesFrom.providerRef is empty or nil,
+	// we interpret it as the data value secret for that package has not been available yet. One of those three fields needs
```

Comment:
Should be a custom error if at all to allow skipping of the package. Lets log an issue to get this use case covered.